### ISSUE TITLE
Crash after: binance responds with "Rest API trading is not enabled" and "market is closed"

### ISSUE THREAD

**Issue by yazeed** [freqtrade/freqtrade#2889]
Today, there was a scheduled maintainence on binance which caused freqtrade to eventually crash.

Suddenly, freqtrade started to print these lines:
```
2020-02-09 03:15:48,904 - freqtrade.strategy.interface - WARNING - Outdated history for pair LTO/BTC. Last tick is 20 minutes old
2020-02-09 03:15:49,913 - freqtrade.strategy.interface - WARNING - Outdated history for pair BEAM/BTC. Last tick is 20 minutes old
```

A while afterwards, it started printing these:
```
2020-02-09 03:17:47,606 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=235, pair=KAVA/BTC, amount=38.86160286, open_rate=0.00011742, open_since=2020-02-09 01:50:42) current_rate=0.00011776
2020-02-09 03:17:49,120 - freqtrade.exchange.common - WARNING - cancel_order() returned exception: "Could not cancel order due to ExchangeNotAvailable. Message: binance Rest API trading is not enabled."
2020-02-09 03:17:49,120 - freqtrade.exchange.common - WARNING - retrying cancel_order() still for 3 times
2020-02-09 03:17:49,624 - freqtrade.exchange.common - WARNING - cancel_order() returned exception: "Could not cancel order due to ExchangeNotAvailable. Message: binance Rest API trading is not enabled."
2020-02-09 03:17:49,624 - freqtrade.exchange.common - WARNING - retrying cancel_order() still for 2 times
2020-02-09 03:17:50,134 - freqtrade.exchange.common - WARNING - cancel_order() returned exception: "Could not cancel order due to ExchangeNotAvailable. Message: binance Rest API trading is not enabled."
2020-02-09 03:17:50,134 - freqtrade.exchange.common - WARNING - retrying cancel_order() still for 1 times
2020-02-09 03:17:50,643 - freqtrade.exchange.common - WARNING - cancel_order() returned exception: "Could not cancel order due to ExchangeNotAvailable. Message: binance Rest API trading is not enabled."
2020-02-09 03:17:50,643 - freqtrade.exchange.common - WARNING - retrying cancel_order() still for 0 times
2020-02-09 03:17:51,198 - freqtrade.exchange.common - WARNING - cancel_order() returned exception: "Could not cancel order due to ExchangeNotAvailable. Message: binance Rest API trading is not enabled."
2020-02-09 03:17:51,198 - freqtrade.exchange.common - WARNING - Giving up retrying: cancel_order()
2020-02-09 03:17:51,198 - freqtrade.worker - WARNING - Error: Could not cancel order due to ExchangeNotAvailable. Message: binance Rest API trading is not enabled., retrying in 30 seconds...
```

And finally, what really crashed it for some reason was when it printed these:
```
2020-02-09 03:23:02,508 - freqtrade.freqtradebot - ERROR - Could not cancel stoploss order 10230038
Traceback (most recent call last):
  File "/Users/yazeed/Sites/freqtradeLATEST/.env/lib/python3.7/site-packages/ccxt/base/exchange.py", line 552, in fetch
    response.raise_for_status()
  File "/Users/yazeed/Sites/freqtradeLATEST/.env/lib/python3.7/site-packages/requests/models.py", line 940, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://api.binance.com/api/v3/order?timestamp=1581214982268&recvWindow=5000&symbol=BEAMBTC&orderId=10230038&signature=asdfadsfasdfasdfasdfadsf

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/yazeed/Sites/freqtradeLATEST/freqtrade/exchange/exchange.py", line 902, in cancel_order
    return self._api.cancel_order(order_id, pair)
  File "/Users/yazeed/Sites/freqtradeLATEST/.env/lib/python3.7/site-packages/ccxt/binance.py", line 1145, in cancel_order
    response = getattr(self, method)(self.extend(request, params))
  File "/Users/yazeed/Sites/freqtradeLATEST/.env/lib/python3.7/site-packages/ccxt/base/exchange.py", line 439, in inner
    return entry(_self, **inner_kwargs)
  File "/Users/yazeed/Sites/freqtradeLATEST/.env/lib/python3.7/site-packages/ccxt/binance.py", line 1676, in request
    response = self.fetch2(path, api, method, params, headers, body)
  File "/Users/yazeed/Sites/freqtradeLATEST/.env/lib/python3.7/site-packages/ccxt/base/exchange.py", line 458, in fetch2
    return self.fetch(request['url'], request['method'], request['headers'], request['body'])
  File "/Users/yazeed/Sites/freqtradeLATEST/.env/lib/python3.7/site-packages/ccxt/base/exchange.py", line 564, in fetch
    self.handle_errors(http_status_code, http_status_text, url, method, headers, http_response, json_response, request_headers, request_body)
  File "/Users/yazeed/Sites/freqtradeLATEST/.env/lib/python3.7/site-packages/ccxt/binance.py", line 1670, in handle_errors
    self.throw_exactly_matched_exception(self.exceptions, error, feedback)
  File "/Users/yazeed/Sites/freqtradeLATEST/.env/lib/python3.7/site-packages/ccxt/base/exchange.py", line 476, in throw_exactly_matched_exception
    raise exact[string](message)
ccxt.base.errors.InvalidOrder: binance {"code":-1013,"msg":"Market is closed."}

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/yazeed/Sites/freqtradeLATEST/freqtrade/freqtradebot.py", line 980, in execute_sell
    self.exchange.cancel_order(trade.stoploss_order_id, trade.pair)
  File "/Users/yazeed/Sites/freqtradeLATEST/freqtrade/exchange/common.py", line 113, in wrapper
    return f(*args, **kwargs)
  File "/Users/yazeed/Sites/freqtradeLATEST/freqtrade/exchange/exchange.py", line 905, in cancel_order
    f'Could not cancel order. Message: {e}') from e
freqtrade.exceptions.InvalidOrderException: Could not cancel order. Message: binance {"code":-1013,"msg":"Market is closed."}
2020-02-09 03:23:03,017 - freqtrade.wallets - INFO - Wallets synced.
2020-02-09 03:23:03,017 - freqtrade.freqtradebot - WARNING - Unable to sell trade: Not enough amount to sell. Trade-amount: 56.62332, Wallet: 0.00448
2020-02-09 03:23:03,513 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=235, pair=KAVA/BTC, amount=38.86160286, open_rate=0.00011742, open_since=2020-02-09 01:50:42) current_rate=0.00011776
2020-02-09 03:23:05,028 - freqtrade.commands.trade_commands - INFO - worker found ... calling exit
2020-02-09 03:23:05,028 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': status, 'status': 'process died'}
2020-02-09 03:23:05,150 - freqtrade.freqtradebot - INFO - Cleaning up modules ...
2020-02-09 03:23:05,150 - freqtrade.rpc.rpc_manager - INFO - Cleaning up rpc modules ...
2020-02-09 03:23:12,373 - freqtrade - ERROR - Could not cancel order. Message: binance {"code":-1013,"msg":"Market is closed."}
```

I thought I'd log this so we can discuss if we should build logic around this to put freqtrade on 1-5min timeouts until the exchange or trading is back up so it wouldn't crash unattended in such cases.

Attempting to rerun freqtrade, crashes it in the same spot.

**Comment by xmatthias** [comment]
I don't think there is a (reliable) endpoint telling us that the exchange is down for maintenance ... so we can't really "wait until it's back up" ... freqtrade simply has to retry until it works again (also considering that there's 2 ways an exchange is down - this (planned) or the complete endpoints are not available (most likely unplanned maintenance).

so I don't mind the bot logging errors until the exchange comes back ... 

----

Now what i don't understand is if the bot is crashing completely - or if it's just logging the exception and you're then cancelling the execution of the bot.

The 2nd error seems to come from `cancel_order()` (cancelling stoploss order) from within `execute_sell()`, and the exception is `InvalidOrderException`.

However, the line in the logs (980 in `freqtradebot.py`) does not call this - but a few lines before that (962) - so i assume you've got some changes earlier in that file (?).

```
  File "/Users/yazeed/Sites/freqtradeLATEST/freqtrade/freqtradebot.py", line 980, in execute_sell
    self.exchange.cancel_order(trade.stoploss_order_id, trade.pair)
```

https://github.com/freqtrade/freqtrade/blob/fff8ced3b059f6901f414a39ea1fe5ed7ff40dee/freqtrade/freqtradebot.py#L960-L966

---

The other error - `ExchangeNotAvailable` is a sub-class of `ccxt.Networkerror` [source](https://github.com/ccxt/ccxt/wiki/Manual#exception-hierarchy) - so that should raise TemporaryError (causing it to retry a few times - then giving up).

Neither of the above errors should stop the bot completely, as far as i can see, they should be handled correctly (unfortunately that's difficult to test - as these exceptions only occur once every few months ...).

**Comment by xmatthias** [comment]
@yazeed any chance you still got the full log of this? (the above seems to miss ~5m) - and it would be interresting to see what happened before the above.

the messages themselfs don't seem to be able to crash the bot when following the stacktrace through the bot ...  
- and sometimes the real crash comes from a few lines earlier - and that exception is just there "by chance" right before it's closing.

**Comment by yazeed** [comment]
Unfortunately, I don't have it anymore.

**Comment by hroff-1902** [comment]
> Maybe it's a good idea to consider having the bot start logging into a folder, with gzip on older log files?

https://www.freqtrade.io/en/latest/advanced-setup/#advanced-logging, the syslog/journald logs are usually rotated and zipped...

**Comment by yazeed** [comment]
Oh, nice, I'll use it.

**Comment by xmatthias** [comment]
> btw, if you follow it, probably you'll find some errors/inconsistences...

in what way? i'm using logging to file since months on 3 bots but i'm not sure what to look at to find these "Inconsistencies".

**Comment by hroff-1902** [comment]
> > btw, if you follow it, probably you'll find some errors/inconsistences...
> 
> in what way? i'm using logging to file since months on 3 bots but i'm not sure what to look at to find these "Inconsistencies" - as nothing does throw errors.

I meant if someone else repeats the steps I wrote in the docs, he could find some inconsistences in the docs I could simply miss in the docs written since the issues were obvious for me...

**Comment by xmatthias** [comment]
ah - i was assuming that inconsistencies would be found in the logs

I've been using --log-file now since several months (in combination with docker ) - but it seems to write and rotate perfectly - so i don't think there's inconsistencies with that (did not try journald though - as i like to keep the logs separate if i have multiple bots running on the same server)

**Comment by xmatthias** [comment]
I think this was closed in #3181 - which fixed the last occurance of cancel_order which did not have the appropriate exception handler.

Feel free to reopen if it reoccurs after updating to the latest version.
